### PR TITLE
Provide TypeScript definition and JSDoc

### LINKED ISSUE #15: Add JSDoc for testRule
For better editor integration (especially VS Code) I would like to add JSDoc to a `testRule`. This way editor would help working with testRule with useful autocomplete.

### PR DESCRIPTION
**Pull request by ybiquitous**
> Which issue, if any, is this issue related to?

Closes #15

Ref: stylelint/stylelint#5725

> Is there anything in the PR that needs further explanation?

This change aims to improve the developer experience when using the `testRule()` utility.

This additionally installs the following package:

```shell
npm i -D typescript @types/jest
```

Also, this supports `endLine` and `endColumn`. See stylelint/stylelint#5725

### To-do

- [x] I'll test if the change actually works in the `stylelint/stylelint` repo.
  - https://github.com/stylelint/stylelint/pull/5932

### CODE REVIEW

File: basicChecks.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+	},
+	{
+		code: '@import "foo.css";',
+		description: 'blockless statement',
+	},
+	{
+		code: ':global {}',
+		description: 'CSS Modules global empty rule set',
+	},
+];
```
Comment (ybiquitous) [inline basicChecks.js:21]: [note] This is a copy from `lib/testUtils/basicChecks.js` in the `stylelint/stylelint` repo.

https://github.com/stylelint/stylelint/blob/42ac6a190c7fdbcc01b35f0642a4111e647c0389/lib/testUtils/basicChecks.js

Within the repo above, `basicChecks.js` seems unused.

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+		skipBasicChecks?: boolean;
+		fix?: boolean;
+		customSyntax?: string | import('postcss').Syntax;
+		codeFilename?: string;
+		only?: boolean;
+		skip?: boolean;
+	};
+
+	/**
+	 * Test a rule with the specified schema.
```
Comment (ybiquitous) [inline index.d.ts:43]: [note] Any feedback to improve the doc is welcome!

**Comment by jeddy3** [comment]
Excellent stuff.

The `basicChecks` stuff is less relevant now that we don't focus on stylistic rules. It's pretty opaque too, and I see it misused when new contributors add new rules.

I suggest we remove the `skipBasicChecks` property as part of this pull request, and publish a major release. What do you think?

**Comment by ybiquitous** [comment]
@jeddy3 Thanks for the background. I agree with removing `skipBasicChecks`. 👍🏼

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+			warnings?: Warning[];
+		};
+
+	type TestSchema = {
+		ruleName: string;
+		config: any;
+		accept?: AcceptTestCase[];
+		reject?: RejectTestCase[];
+		plugins?: string | string[];
+		skipBasicChecks?: boolean;
```
Comment (jeddy3) [inline index.d.ts:34]:
```suggestion
```

Shall we remove it from the schema too?

File: index.d.ts (hunk 2 above)
Comment (jeddy3) [reply index.d.ts:43]: LGTM as it is.

File: index.d.ts (hunk 3 above)
Comment (ybiquitous) [reply index.d.ts:34]: Thanks. Removed. 👍🏼

File: README.md [hunk 4]
```diff
@@ -210,7 +210,7 @@ Don't check the `fixed` code.
 
 Optional if `message` is used.
 
-Warning Objects containing expected `message`, `line` and `column`.
+Warning objects containing expected `message`, `line` and `column`.
 
 ## [Changelog](CHANGELOG.md)
```
[the remark this answers is not in the document]
Comment (jeddy3) [reply README.md:216]: LGTM. Let's just maintain the type defs and link to them like you are.

[the remark this answers is not in the document]
Comment (ybiquitous) [reply README.md:112]: Thanks. Changed via 5509654.

**Review by jeddy3 — APPROVED** [review]
Fantastic!

LGTM.

Manually tested with [`stylelint-scales`](https://github.com/signal-noise/stylelint-scales) (a plugin pack I maintain).

**Comment by jeddy3** [comment]
Shall we add three changelog entries after merge?:

- Removed: `skipBasicChecks` schema property.
- Added: `endLine` and `endColumn` schema properties.
- Added: TypeScript definitions.

**Comment by ybiquitous** [comment]
Can you please merge, update the changelog, and release a new version if possible?

**Comment by ybiquitous** [comment]
Wait, I find a typo. I try fixing it.

**Comment by jeddy3** [comment]
> Can you please merge, update the changelog, and release a new version if possible?

Can do.

**Comment by jeddy3** [comment]
Released in 5.0.0

**Comment by ybiquitous** [comment]
Thank you so much! 🎉

**Pull request merged**